File: day4.py
class Board:
    def __init__(self, values: list[list[int]]):
        self.__values = values

    # Replace every occurance of 'num' with -1
    def draw(self, num: int) -> None:
        height = len(self.__values)
        width = len(self.__values[0])
        for y in range(height):
            for x in range(width):
                if self.__values[y][x] == num:
                    self.__values[y][x] = -1

    # Check if -1's form a horizontal or vertical line
    def hasWon(self) -> bool:
        height = len(self.__values)
        width = len(self.__values[0])

        # Check horizontal lines
        for y in range(height):
            won = True
            for x in range(width):
                if self.__values[y][x] != -1:
                    won = False
                    break
            if won:
                return True

        # Check vertical lines
        for x in range(width):
            won = True
            for y in range(height):
                if self.__values[y][x] != -1:
                    won = False
                    break
            if won:
                return True
        
        return False

    # Calculate the sum of non -1 values
    def getScore(self) -> int:
        sum = 0
        for line in self.__values:
            for value in line:
                if value != -1:
                    sum += value
        return sum

    # Print the board's values
    def print(self):
        for line in self.__values:
            print(line)

def ParseFile(path: str):
    draws = list[int]()
    boards = list[Board]()

    with open(path) as file:
        lines = file.readlines()

        # Draws are comma seperated list on the first line
        draws = [int(x) for x in lines[0].strip().split(',')]

        # Parse and add boards to 'boards'
        values = list[list[int]]()
        for i in range(1, len(lines)):
            line = lines[i].strip()

            # On empty line add the current board and reset 'values'
            # otherwise append the current line to 'values'
            if len(line) == 0: 
                if len(values) > 0:
                    boards.append(Board(values))
                    values = list[list[int]]()
            else:
                values.append([int(x) for x in line.split()])

        # Add the last board if the file did not end with newline
        if len(values) > 0:
            boards.append(Board(values))

    return draws, boards

def part1(path: str) -> int:
    draws, boards = ParseFile(path)

    for num in draws:
        for board in boards:
            board.draw(num)
            if board.hasWon():
                return num * board.getScore()

def part2(path: str) -> int:
    draws, boards = ParseFile(path)

    for num in draws:
        for board in boards:
            board.draw(num)
        if len(boards) == 1 and boards[0].hasWon():
            return num * boards[0].getScore()
        boards = [x for x in boards if not x.hasWon()]

if __name__ == '__main__':
    print(f"PART 1: {part1('input/day4.txt')}")
    print(f"PART 2: {part2('input/day4.txt')}")

File: day4.cpp
#include "utilities.h"

struct Board
{
    int cells[5][5]{};

    Board(std::ifstream& stream)
    {
        std::string word;
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (!(stream >> word)) {
                    cells[0][0] = -1;
                    return;
                }
                cells[i][j] = std::stoi(word);
            }
        }
    }

    bool Wins()
    {
        for (int i = 0; i < 5; i++) {
            bool h = true;
            bool v = true;
            for (int j = 0; j < 5; j++) {
                if (cells[i][j] != -1)
                    h = false;
                if (cells[j][i] != -1)
                    v = false;
            }
            if (h || v) return true;
        }
        return false;
    }

    void Draw(int val)
    {
        for (int i = 0; i < 5; i++)
            for (int j = 0; j < 5; j++)
                if (cells[i][j] == val)
                    cells[i][j] = -1;
    }

    int GetSum()
    {
        int sum = 0;
        for (int i = 0; i < 5; i++)
            for (int j = 0; j < 5; j++)
                if (cells[i][j] != -1)
                    sum += cells[i][j];
        return sum;
    }

};

std::vector<Board> LoadBoards(std::ifstream& stream)
{
    std::vector<Board> boards;
    while(true)
    {
        Board board(stream);
        if(board.cells[0][0] == -1)
            break;
        boards.push_back(board);
    }
    return boards;
}

int part1(char* path)
{
    std::vector<int> draws;
    std::vector<Board> boards;

    {
        std::ifstream file(path);
        draws = ParseCSV(file);
        boards = LoadBoards(file);
    }
    
    for (int num : draws) {
        for (Board& board : boards) {
            board.Draw(num);
            if (board.Wins())
                return board.GetSum() * num;
        }
    }
    
    return -1;
}

int part2(char* path)
{
    std::vector<int> draws;
    std::vector<Board> boards;

    {
        std::ifstream file(path);
        draws = ParseCSV(file);
        boards = LoadBoards(file);
    }

    for (int num : draws) {
        for (auto it = boards.begin(); it != boards.end();) {
            it->Draw(num);
            if (it->Wins()) {
                if (boards.size() == 1)
                    return it->GetSum() * num;
                it = boards.erase(it);
            } else it++;
        }
    }

    return -1;
}

int main(int argc, char** argv)
{
    char* path;

    if(argc != 2)
    {
        path = new char[100];
        sprintf(path, "input/%s.txt", argv[0]);
    }
    else path = argv[1];

    printf("PART 1: %i\n", part1(path));
    printf("PART 2: %i\n", part2(path));

    if(argc != 2)
        delete[] path;
}